### PR TITLE
[wasm] Allow to build browser in library mode

### PR DESCRIPTION
**Pull request by mkhamoyan**
By this PR we are supporting library mode for browser.

Contributes to https://github.com/dotnet/runtime/issues/77191

### CODE REVIEW

File: src/tools/illink/src/ILLink.Tasks/build/Microsoft.NET.ILLink.targets [hunk 1]
```diff
@@ -140,6 +140,11 @@ Copyright (c) .NET Foundation. All rights reserved.
     </PropertyGroup>
 
     <Delete Files="@(_LinkedResolvedFileToPublishCandidate)" />
+    <ItemGroup>
+      <TrimmerRootAssembly Condition="'$(RuntimeIdentifier)' == 'browser-wasm' And '%(Identity)' == '@(IntermediateAssembly)' And '$(OutputType)' == 'Library'">
+        <RootMode>all</RootMode>
+      </TrimmerRootAssembly>
+    </ItemGroup>
```
Comment (mkhamoyan) [inline src/tools/illink/src/ILLink.Tasks/build/Microsoft.NET.ILLink.targets:147]: Not sure this is the right place.

File: src/tasks/WasmAppBuilder/PInvokeTableGenerator.cs [hunk 2]
```diff
@@ -386,11 +386,12 @@ private void EmitNativeToInterp(StreamWriter w, List<PInvokeCallback> callbacks)
             sb.Append(") { \n");
             if (!is_void)
                 sb.Append($"  {MapType(method.ReturnType)} res;\n");
-
+#if TARGET_WASI
```
[the remark this answers is not in the document]
Comment (mkhamoyan) [reply src/tasks/WasmAppBuilder/PInvokeTableGenerator.cs:389]: I couldn't find RID passed to the task https://github.com/dotnet/runtime/blob/67df47e1abb45b08f32063737e1e0f6eecb083e7/src/mono/wasi/wasi.proj#L65-L72  that's why decided to check  `TARGET_WASI`.

[the remark this answers is not in the document]
Comment (maraf) [reply src/tasks/WasmAppBuilder/PInvokeTableGenerator.cs:389]: Yeah, we need to pass-in RID. `TARGET_WASI` wouldn't solve the problem

**Review by lewing — CHANGES_REQUESTED** [review]
Needs follow up on existing comments

File: src/tasks/WasmAppBuilder/PInvokeTableGenerator.cs [hunk 3]
```diff
@@ -26,13 +26,15 @@ internal sealed class PInvokeTableGenerator
     private readonly List<PInvokeCallback> callbacks = new();
     private readonly PInvokeCollector _pinvokeCollector;
     private readonly bool _isLibraryMode;
+    private readonly string _runtimeIdentifier;
 
-    public PInvokeTableGenerator(Func<string, string> fixupSymbolName, LogAdapter log, bool isLibraryMode = false)
+    public PInvokeTableGenerator(Func<string, string> fixupSymbolName, LogAdapter log, bool isLibraryMode = false, string runtimeIdentifier = "browser-wasm")
```
Comment (maraf) [inline src/tasks/WasmAppBuilder/PInvokeTableGenerator.cs:31]: nit: Please don't fallback to browser, make the argument "required"

**Review by maraf — APPROVED** [review]
Looks good to me 👍

**Review by ilonatommy — APPROVED** [review]

File: src/tasks/WasmAppBuilder/ManagedToNativeGenerator.cs [hunk 4]
```diff
@@ -35,6 +35,8 @@ public class ManagedToNativeGenerator : Task
 
     public bool IsLibraryMode { get; set; }
 
+    public string RuntimeIdentifier { get; set; } = string.Empty;
```
[the remark this answers is not in the document]
Comment (mkhamoyan) [reply src/tasks/WasmAppBuilder/ManagedToNativeGenerator.cs]: Thanks, updated.

**Comment by lewing** [comment]
This feels wrong, we're plumbing the rid down to the generator rather than dealing with runtime init ordering in the browser case.  I'd rather we unconditionally emit the intiialize_runtime call and have it assert when it can't actually do the initialization.

**Comment by maraf** [comment]
> This feels wrong, we're plumbing the rid down to the generator rather than dealing with runtime init ordering in the browser case. I'd rather we unconditionally emit the intiialize_runtime call and have it assert when it can't actually do the initialization.

We are initializing runtime for browser in the place™️ manually. I guess we can have wrap the `mono_wasm_load_runtime` in `initialize_runtime` for browser case, still call it manually and leave the generator untouched

**Review by lewing — APPROVED** [review]

**Pull request merged**